.github/workflows/evaluation-test-reusable.yaml

```diff
@@ -0,0 +1,193 @@
+
+      - name: K3d Setup - Provision K3d cluster
+        run: |
+          echo "::group::K3d - Provision"
+          k3d cluster create test-cluster --verbose --agents=1 -p 9080:80@loadbalancer -p 9443:443@loadbalancer -p "32000:32000@agent:0" -p "32100:32100@agent:0"
+          echo "::endgroup::"
+
+      - name: K3d Setup - Verify K3d context
+        run: |
+          echo "::group::K3d - Context"
+          context_name=$(kubectl config current-context)
+          if [ "$context_name" != "k3d-test-cluster" ]; then
+            echo "K3d context is not correct. Expected: k3d-test-cluster, got: $context_name"
+            exit 1
+          fi
+          echo "K3d context is correct"
+          echo "::endgroup::"
+
+      - name: Redis Setup - Deploy Redis using Helm
+        run: |
+          kubectl create namespace redis
+          helm repo add redis-stack https://redis-stack.github.io/helm-redis-stack/
+          echo "Installing Redis using Helm..."
+          helm install --wait redis-stack redis-stack/redis-stack --set auth.enabled=false -n redis
+
+      - name: Redis Setup - Change NodePort
+        run: |
+          kubectl patch service redis-stack -n redis --type='json' -p='[{"op": "replace", "path": "/spec/ports/0/nodePort", "value": 32100}]'
+          kubectl get svc -n redis redis-stack -o yaml
+
+      - name: Companion Deploy - Create secret
+        env:
+          COMPANION_CONFIG_BASE64: ${{ secrets.EVALUATION_TESTS_CONFIG }}
+        run: |
+          kubectl create namespace ai-system
+          ./scripts/k8s/create-secret.sh
+
+      - name: Companion Deploy - Apply companion manifests
+        run: |
+          kubectl apply -f scripts/k8s/companion-k3d-manifest.yaml
+
+      - name: Companion Deploy - Update image name in Deployment
+        run: |
+          echo "Updating companion image to ${IMAGE_NAME}..."
+          kubectl -n ai-system set image deployment/companion companion=${IMAGE_NAME}
+
+      - name: Companion Deploy - Wait for deployment
+        run: |
+          kubectl rollout status deployment/companion -n ai-system --timeout=300s
+
+      - name: Companion Deploy - Test reachability through NodePort
+        env:
+          TIMEOUT_SECONDS: 300  # Default timeout (5 minutes)
+          INTERVAL_SECONDS: 10  # Default interval (10 seconds)
+        run: |
+          end_time=$((SECONDS + TIMEOUT_SECONDS))
+          while [ $SECONDS -lt $end_time ]; do
+            if curl -s -o /dev/null -w "%{http_code}" http://localhost:32000/readyz | grep -q 200; then
+              echo "Service is reachable";
+              exit 0;
+            fi
+            echo "Service not reachable yet, retrying in $INTERVAL_SECONDS seconds...";
+            sleep $INTERVAL_SECONDS;
+          done
+          echo "Service did not become reachable within $TIMEOUT_SECONDS seconds";
+          exit 1
+
+      - name: Companion Deploy - Debug information
+        if: failure()
```

Comment:
let run this step for both cases, success and failure. To have audit logs.

Reply:
I see that we do this twice. I do not see why so I removed one but made the second one, at the bottom of the flow, run always.

Reply:
We have this twice in the code so I changed it for both places.